Add profiles for imv, retroarch, and torbrowser

crocket opened the pull request:
The profiles are based on profile template.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+# Persistent local customizations
```
Headers

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+include disable-devel.inc
+include disable-exec.inc
+include disable-interpreters.inc
+include disable-passwdmgr.inc
```
```suggestion
```

crocket replied:
Why remove disable-passwdmgr?

kmk3 replied:
@crocket [on Oct 9](https://github.com/netblue30/firejail/pull/4600#discussion_r725546759):

> Why remove disable-passwdmgr?

It does not exist anymore.

See #4461 and the current profile.template.

crocket replied:
Removed from my local copy.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+include disable-shell.inc
+include disable-write-mnt.inc
+
+include whitelist-runuser-common.inc
```
Can we use wvc, wrc, wusc?

crocket replied:
I thought I wanted to view images in var and usr.

rusty-snake replied:
usr/share makes sense. var, IDK if it contains any images except flatpak icons, run can be added imho.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+caps.drop all
+net none
+nodvd
+noinput
```
`nogroups`?

crocket replied:
After further testing, it turned out that what broke imv was actually file permission on images.

Some image files were not owned by the normal user. `nogroup` and `noroot` don't actually break imv.

Added it to my local copy.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+nodvd
+noinput
+nonewprivs
+nosound
```
`noroot`?

crocket replied:
noroot breaks imv on my gentoo linux system.

crocket replied:
After further testing, it turned out that what broke imv was actually file permission on images.

Some image files were not owned by the normal user. `nogroup` and `noroot` don't actually break imv.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+notv
+nou2f
+novideo
+protocol unix
```
Does it need `unix`?

crocket replied:
It seems that unix socket is required for wayland.

rusty-snake replied:
> Can net none alone disable unix sockets?

No it is not a syscall related option, it is a namespace related option. To block all protocols, use `seccomp socket`. Anyway if it uses wayland ...

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+nou2f
+novideo
+protocol unix
+seccomp
```
`seccomp.block-secondary`?

crocket replied:
Done

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+novideo
+protocol unix
+seccomp
+shell none
```
`tracelog`?

crocket replied:
Why?

rusty-snake replied:
Because we add it to every profile where it works.

rusty-snake commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,40 @@
+seccomp
+shell none
+
+private-bin imv,imv-wayland,imv-x11,sh
```
`private-cache`?

crocket replied:
Added it to my local copy.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+# Persistent local customizations
```
Headers

crocket replied:
Added headers to my local copy.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+# Persistent global definitions
+include globals.local
+
+blacklist /boot
+blacklist /opt
```
```suggestion
```

rusty-snake replied:
`/boot` is a hardcoded blacklist, you don't need to repeat it.

`/opt` might me necessary if someone installs it there ...

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+blacklist /boot
+blacklist /opt
+blacklist /usr/libexec
+blacklist /var
```
```suggestion
```

crocket replied:
It works fine without /var.

rusty-snake replied:
IDC if it works fine for you without `/var` it will not work on systems which use `/var/lib/ca-certificates` for example.

crocket replied:
Just removed it.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+include disable-passwdmgr.inc
+include disable-programs.inc
+include disable-shell.inc
+include disable-write-mnt.inc
```
```suggestion
```

crocket replied:
Okay, people may put their ROMs in /mnt. Removed it from my local copy.

rusty-snake replied:
No.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+include disable-devel.inc
+include disable-exec.inc
+include disable-interpreters.inc
+include disable-passwdmgr.inc
```
```suggestion
```

crocket replied:
???

crocket replied:
Removed from my local copy.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+whitelist ${HOME}/.config/retroarch
+whitelist /usr/share/retroarch
+whitelist /usr/share/libretro
+
```
```suggestion
```

crocket replied:
Done in my local copy.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+include disable-write-mnt.inc
+include disable-xdg.inc
+
+whitelist ${HOME}/.config/retroarch
```
`mkdir`+`blacklist`(in dp.inc)+`noblacklist`

crocket replied:
????

rusty-snake replied:
https://github.com/netblue30/firejail/blob/07e9029b24e21de4c6ecae3060bed3d2450bb296/etc/templates/profile.template#L62-L64

```suggestion
mkdir ${HOME}/.config/retroarch
whitelist ${HOME}/.config/retroarch
```

crocket replied:
I don't understand why `whitelist` requires `mkdir`, but I just added it.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+
+include whitelist-common.inc
+include whitelist-runuser-common.inc
+include whitelist-usr-share-common.inc
```
Can we use wr and wvc?

crocket replied:
Why?

rusty-snake replied:
Why not?

crocket replied:
It is fine with

```
include whitelist-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
include whitelist-var-common.inc
```

rusty-snake replied:
and wrc?

crocket replied:
wrc is `include whitelist-runuser-common.inc`.

rusty-snake replied:
Review comment says

> wrc: whitelist-run-common.inc

and whitelist-runuser-common.inc is wruc

crocket replied:
Yes, I applied it to retroarch.profile.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+nou2f
+protocol unix,inet,inet6
+seccomp
+shell none
```
`tracelog`?

rusty-snake replied:
Only if you blacklist a file that it tries to access in a loop.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+notv
+nou2f
+protocol unix,inet,inet6
+seccomp
```
`seccomp.block-secondary`?

crocket replied:
I'm going to test it with every libretro core.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+apparmor
+caps.drop all
+netfilter
+nodvd
```
`nogroups`?

crocket replied:
nogroups breaks my gentoo linux system.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+netfilter
+nodvd
+nonewprivs
+notv
```
`noroot`?

crocket replied:
noroot breaks my gentoo linux system.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+nodvd
+nonewprivs
+notv
+nou2f
```
`novideo`?

crocket replied:
I don't know what exactly `novideo` does. What are video devices?

crocket replied:
It seems `novideo` turns off cameras. I added it to my local copy.

rusty-snake replied:
IDK how common/necessary it is. Maybe add a comment.

crocket replied:
Added with a comment.

rusty-snake commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,46 @@
+shell none
+
+disable-mnt
+private-bin retroarch
```
`private-cache`?

crocket replied:
Added it to my local copy.

rusty-snake commented on etc/profile-m-z/torbrowser.profile:
```diff
@@ -0,0 +1,22 @@
+# Persistent local customizations
+include torbrowser.local
+# Persistent global definitions
+include globals.local
+
+ignore dbus-user none
+
+noblacklist ${HOME}/.cache/mozilla
+noblacklist ${HOME}/.mozilla
+
+blacklist /usr/libexec
+
+mkdir ${HOME}/.cache/mozilla/torbrowser
+mkdir ${HOME}/.mozilla
+whitelist ${HOME}/.cache/mozilla/torbrowser
+whitelist ${HOME}/.mozilla
+include whitelist-usr-share-common.inc
+
+dbus-user filter
+dbus-user.own org.mozilla.torbrowser.*
+
+include firefox-common.profile
```
```suggestion
# Firejail profile for torbrowser
# This file is overwritten after every install/update
# Persistent local customizations
include torbrowser.local
# Persistent global definitions
# added by included profile
#include globals.local

# Redirect
include start-tor-browser.profile
```

crocket replied:
www-client/torbrowser::torbrowser on Gentoo Linux doesn't use torbrowser-launcher. It compiles torbrowser directly from source code.

rusty-snake replied:
Ahh, it is only for the gentoo package, then add this information to the description.

crocket replied:
I just did in my local copy. It will be reflected soon in the pull request.

kmk3 commented:
@crocket commented [on Oct 9](https://github.com/netblue30/firejail/pull/4600#issuecomment-939374645):

> Whenever I read `nogroups` and `noroot`, I feel marginalized as someone who
> doesn't use systemd.
> 
> Quite a few linux distributions don't use systemd by default.
> 
> * Void Linux
> * Gentoo Linux
> * Artix Linux
> * GNU Guix System
> * ...
> 
> https://nosystemd.org/
> 
> Don't break linux distributions for systemd.

What do you mean?  I use Artix Linux myself and I don't remember having
problems with either option.

crocket commented:
I tell you. Either `nogroup` or `noroot` breaks imv and retroarch.

rusty-snake commented:
> Whenever I read nogroups and noroot, I feel marginalized as someone who doesn't use systemd.
> Don't break linux distributions for systemd.

Users who refuse the technical progress of the last ten years are no reason to reduce the security for all. 

> Quite a few linux distributions don't use systemd by default.

And here's their distrowatch ranking, guess what the majority of users have.

1. MX Linux
3. Manjaro
4. Mint
6. Ubuntu
7. Debian
10. Fedora
12. openSUSE
16. Arch
19. Artix
35. Devuan
41. Void
53. Gentoo

> https://nosystemd.org/

Sorry, but this site is just bullshit.

> List of notable bugs and security issues:

Guess where bugs are found, it the software used by a few  people with a heavily selfconfigured system or in software used by the majority of users and in all enterprise distribution.

> systemd will be able to kill a process if he needs/wants more RAM

Sorry but that's a feature.

> systemd does not respect system wide resource limits

TBH, resource limit configuration is a mess like shell startup configuration. And `/etc/security/limits.conf` isn't "the system-wide resource limits" configuration file. RT(F)M "configuration file for the pam_limits module". If something does not use pam at all of a pam-configuraton without pam_limits ...

crocket commented:
> In a multi-user system, a remotely logged in user can listen to microphone, camera and input devices when another user is logged in locally. The users can also make set-gid binaries to keep access even if the group would be removed.

That particular attack scenario is rare and difficult to pull off. There is no way that someone is just remotely logging in my computer. Don't assume that remote login is easy. Even after remote login, making a setgid binary as a non-root user isn't trivial. There is no easy shortcut.

Not sure about `nogroups`, but `noroot` is questionable from security standpoint. It uses user namespace which may have more security holes than an SUID binary such as firejail.

> Users who refuse the technical progress of the last ten years are no reason to reduce the security for all.

Some people say systemd represents technical progress, but I don't see much progress being made by systemd.

It doesn't offer enough real advantages over other alternatives to entice me. It is just a big lump of softwares that do a lot of things. Technical progress isn't measured by the scope or size of a software. I used to use systemd on Arch Linux. But, I settled with OpenRC because it strikes a good balance between features and simplicity. OpenRC does what I want without much hassle and without sysvinit.

We haven't had much technical progress in general over the last ten years. CPU speed hasn't increased much. Software quality certainly hasn't increased. Consider the possibility that there hasn't been much technical progress in general over the last 10 years.

I also think tools like firejail shouldn't be too dependent on one specific software such as systemd. Detecting systemd with `NOSYSTEMD?` is like a web page detecting browsers. Nowadays, web pages detect browser capabilities.

To make firejail work smoothly on all linux distributions and perhaps BSDs, firejail should detect capabilities.

topimiettinen commented:
> > In a multi-user system, a remotely logged in user can listen to microphone, camera and input devices when another user is logged in locally. The users can also make set-gid binaries to keep access even if the group would be removed.
> 
> Actual multi-user system doesn't exist outside shared hosting. Almost everyone has dedicated machines. That particular attack scenario is rare and difficult to pull off. There is no way that someone is just remotely logging in my computer. Don't assume that remote login is easy. Even after remote login, making a setgid binary as a non-root user isn't trivial. There is no easy shortcut.

Multi-user (even multi-seat) systems exist in internet cafés. The totally unprivileged command is `chmod g+s filename`. It's really trivial and easy.

Do you use `elogind`?

crocket commented:
Internet cafes don't use linux. Almost all of them use windows. Also, internet cafe computers are used by one person at a time.

Firejail targets home market for people who are sensitive to security and privacy.  It's safe to assume that firejail users know how to secure their systems.

topimiettinen commented:
> Internet cafes don't use linux. Almost all of them use windows. Forget about security in internet cafe computers. It's a lost cause. You are not going to get it. If you don't manage the computer that you are using, don't expect it to be secure and private.

Then there are schools and libraries. The system administrator must of course be trusted, but it's another level when you can't trust other unprivileged users either.

> Restricted access to supplementary groups has yet to be implemented.

Restricted access to devices has been implemented, it's called `uaccess` and ACLs. elogind also implements this.

I think you didn't get my idea of `?NOSYSTEMD`. The configurations would remain exactly same for systemd/elogind and other users. For systemd/elogind users, `?NOSYSTEMD ignore noroot` (or perhaps`?SYSTEMD noroot`) would enable improved security of `noroot`. Other users would also get a working system with the same configuration, but without the extra security. The magic directive `?NOSYSTEMD` would do the work.

> I replaced elogind with seatd

It seems that seatd doesn't implement `uaccess`/ACL stuff. It would probably not be too difficult to add it [here](https://git.sr.ht/~kennylevinsen/seatd/tree/master/item/seatd/seat.c).

crocket commented:
schools and libraries also use windows or sometimes Mac OS. Linux is only dominant in server market. I'm not counting android as linux.

I think `?NOSYSTEMD` is inaccurate. As far as I know, systemd alone can't give temporary access to audio/input devices.

topimiettinen commented:
> schools and libraries also use windows or sometimes Mac OS. Linux is only dominant in server market. I'm not counting android as linux.

Wikipedia entry for [Multiseat configuration](https://en.wikipedia.org/wiki/Multiseat_configuration) lists installations in Brazil and Tanzania. There's also a [company](http://www.multi-seat.com/) selling multiseat equipment.

> I think `?NOSYSTEMD` is inaccurate. As far as I know, systemd alone can't give temporary access to audio/input devices. It's the job of pulseaudio or pipewire.

Right, choice of PID1 doesn't have anything to do with this. It's a combination of udevd and systemd-logind/elogind. `?UACCESS` would be more accurate, though a bit cryptic. 

Actually it would be simple to implement this as a `configure` option, for example `--without-systemd` or `--without-uaccess`:
- When Firejail is built with `configure --without-systemd`, `?UACCESS` would drop the rest of the commands. I suppose this could be enabled in Gentoo with USE flags which exclude both systemd and elogind.
- When Firejail is built with  `configure --with-systemd` (the default), `?UACCESS` would execute the commands. In Gentoo, USE flags systemd or elogind would enable this.

I don't know too much about Gentoo (the one time I tried to install it, the kernel which took ages to compile didn't boot), so the examples could be wrong.

rusty-snake commented:
> schools and libraries also use windows or sometimes Mac OS. Linux is only dominant in server market.

https://www.linuxmuster.net

topimiettinen commented:
> We already disable `nogroups` for nVidia users at runtime maybe this code could be extended.

I don't know how to detect absence of systemd-logind or elogind at runtime, except maybe by using APIs of `libsystemd` (linking it to Firejail). The nice thing about the build time switch is that anti-systemd distros can get what they want (their users like @crocket could contribute nice profiles that can be useful elsewhere too) but it doesn't reduce security for modern distros. Perhaps it wouldn't work correctly in the case where systemd is installed, but I think in that case Gentoo users could be expected to recompile all things enabling systemd support everywhere anyway and other distros may have separate systemd/non-systemd package variants or whatever.

I think nvidia driver can't use uaccess because of some GPL issues or limitations to APIs/ABIs for non-GPL drivers. Or maybe the driver is just written to work with the lowest common denominator.

crocket commented:
`noroot` or `nogroups` will break retroarch.profile.
For now, `noroot` and `nogroups` can reside without `?UACCESS` in torbrowser.profile.
`noroot` and `nogroups` didn't break imv.profile.

crocket commented on etc/profile-a-l/imv.profile:
```diff
@@ -0,0 +1,52 @@
+private-tmp
+
+dbus-user none
+dbus-system none
```
`private-tmp` makes `read-only /tmp` redundant.

rusty-snake replied:
Not necessarily.

crocket replied:
By default, `read-only ${HOME}` and `read-only /tmp` don't break imv. But, customized user scripts executed from imv may want to write to ${HOME}. I added `read-only ${HOME}` and `read-only /tmp` to my local copy.

Is `read-only /tmp` useful in the presence of `private-tmp`?

davidebeatrici commented on etc/profile-m-z/retroarch.profile:
```diff
@@ -0,0 +1,52 @@
+caps.drop all
+netfilter
+nodvd
+# nogroups breaks udev input driver which requires access to /dev/input
+#nogroups
+nonewprivs
+# noroot breaks udev input driver which requires access to /dev/input
+#noroot
```
@crocket How do the permissions in `/dev/input` look like on your system?

crocket replied:
I just applied noroot and nogroups to retroarch.profile and pushed the change. Can somebody test retroarch.profile on a system that has (e)logind?

crocket commented:
I just added a comment to imv.profile. I also added disable-mnt to retroarch.profile because `man hier` says `/mnt` is for temporarily mounted filesystems.

crocket commented:
I guess the profile already works on logind systems. I suggest just merging this pull request and deal with any problems later.

crocket commented:
I just noticed that sdl2 controller driver doesn't work inside firejail. It needs a little more work.

/var/log/messages

```
Oct 17 21:35:59 localhost kernel: retroarch[14465]: segfault at 28 ip 00007fc6332f9c2a sp 00007ffcb6bf61f8 error 4 in libSDL2-2.0.so.0.14.0[7fc633214000+e9000]
```

crocket commented:
I just fixed it by allowing netlink protocol which is somehow needed by sdl2.
